lib/internal/http2/core.js

```diff
@@ -307,8 +307,20 @@ function onStreamClose(code) {
 
   if (state.fd !== undefined)
     tryClose(state.fd);
-  stream.push(null);
-  stream[kMaybeDestroy](null, code);
+
+  // defer this until we actually emit end
+  if (stream._readableState.endEmitted || code !== NGHTTP2_NO_ERROR) {
+    stream[kMaybeDestroy](null, code);
+  } else {
+    stream.on('end', stream[kMaybeDestroy]);
+    stream.push(null);
+
+    // TODO mcollina: are we sure this is correct?
+    if (stream.readableLength === 0) {
+      // autoresume, the stream needs to finish
+      stream.resume()
```

Comment:
What `net.Socket`s do is calling `.read(0)` unconditionally after `.push(null)`. Could you try to see if that makes sense? I would really like to merge these bits of code and it would be great if we could avoid these discrepancies…